src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6338,120 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                                break;
+
+                            case SyntaxKind.RefKeyword:
+                            case SyntaxKind.OutKeyword:
+                            case SyntaxKind.InKeyword:
+                                modifiers.Add(EatToken(CurrentToken.Kind));
+                                break;
+                        }
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.FunctionPointerParameterOrReturnType(modifiers, parameterType));
+
+                        if (CurrentToken.Kind == SyntaxKind.CommaToken)
+                        {
+                            types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                        }
+                        else if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        PostSkipAction skipBadFunctionPointerParameterListTokens()
+                        {
+                            CSharpSyntaxNode tmp = null;
+                            Debug.Assert(types.Count > 0);
+                            return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                                types,
+                                isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                                abortFunction: p => p.CurrentToken.Kind == SyntaxKind.CommaToken || p.IsTerminator(),
+                                expected: SyntaxKind.CommaToken);
+                        }
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            SyntaxToken parseInvalidCallingConvention()
+            {
+                // If the user typed something other than a simple identifier
+                // (such as a valid name or a predefined type) in this spot,
+                // then assume that it's not an invalid calling convention.
+                // We'll get an unexpected token error when the next step
+                // attempts to a eat a <
+                if (CurrentToken.Kind != SyntaxKind.IdentifierToken ||
+                    SyntaxFacts.IsPredefinedType(CurrentToken.Kind))
+                {
+                    return null;
+                }
```

Comment:
note: i think we should scan and be resilient to people trying to write things like so:

```c#
delegate* ret_type<arg_types> // and
delegate* ret_type(arg_types) // and
delegate* ret_type(parameters) // and
```

We should detect this and give good errors on what it actually expected.  This is similar to how we detect common reorderings of operator pieces since people commonly confuse them.  

Note: i think it's arguable that your syntax model should *actually* be:

```delegate * calling_conv? < parameters >```

You can then validate *later* that the parameters don't have names/initializers.  But it will mean we can reuse existing parsing logic, and it means that the syntax model can accurately represent common mistakes that people *will make*.

Reply:
I've added test cases for all of these scenarios, though the `ret_type` part is actually interpreted as an invalid calling convention rather than a return type.